Microsoft.Toolkit.Uwp.UI/AdvancedCollectionView/AdvancedCollectionView.cs

```diff
@@ -642,50 +639,72 @@ private void SourceNcc_CollectionChanged(object sender, NotifyCollectionChangedE
             }
         }
 
-        private void HandleItemAdded(int newStartingIndex, object newItem)
+        private bool HandleItemAdded(int newStartingIndex, object newItem, int? viewIndex = null)
         {
             if (_filter != null && !_filter(newItem))
             {
-                return;
+                return false;
             }
 
+            var newViewIndex = _view.Count;
+
             if (_sortDescriptions.Any())
             {
                 _sortProperties.Clear();
-                newStartingIndex = _view.BinarySearch(newItem, this);
-                if (newStartingIndex < 0)
+                newViewIndex = _view.BinarySearch(newItem, this);
+                if (newViewIndex < 0)
                 {
-                    newStartingIndex = ~newStartingIndex;
+                    newViewIndex = ~newViewIndex;
                 }
             }
             else if (_filter != null)
             {
                 if (_sourceList == null)
                 {
                     HandleSourceChanged();
-                    return;
+                    return false;
                 }
 
-                var visibleBelowIndex = 0;
-                for (var i = newStartingIndex; i < _sourceList.Count; i++)
+                if (newStartingIndex == 0 || _view.Count == 0)
+                {
+                    newViewIndex = 0;
+                }
+                else if (newStartingIndex == _sourceList.Count - 1)
                 {
-                    if (!_filter(_sourceList[i]))
+                    newViewIndex = _view.Count - 1;
+                }
+                else if (viewIndex.HasValue)
+                {
+                    newViewIndex = viewIndex.Value;
+                }
+                else
+                {
+                    var viewSet = new Dictionary<object, int>(_view.Count);
```

Comment:
This block should perform better and work better with lists containing duplicate elements `{ "A", "B", "C", "A", "C", "A" }` like so:
```csharp
for (int i = 0, j = 0; i < _sourceList.Count; i++)
{
    if (i == newStartingIndex)
    {
        newViewIndex = j;
        break;
    }
    if (_view[j] == _sourceList[i])
    {
        j++;
    }
}
```

Reply:
Great point, I like this solution a lot more.  Much better solution than creating a Dict like I am.  Gracias!